dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -335,15 +364,68 @@ def read_filepath(path):
         self._cache_outputs = dict()
         # Whether we want the order of DALI execution to be reproducible
         self._deterministic = deterministic
+        # Proxy
+        self._proxy = None
+
+    def _check_dali_input_names(self, input_names):
+        pipe_input_names = _external_source_node_names(self._pipe)
+        if len(pipe_input_names) == 0:
+            raise RuntimeError("The provided pipeline doesn't have any inputs")
+        pipe_input_names_set = set(pipe_input_names)
+        input_names_set = set(input_names or [])
+        if len(input_names_set) != len(input_names_set):
+            raise RuntimeError("``input_names`` argument should not contain any duplicated values")
+
+        if len(input_names_set) == 0:
+            allow_positional_args = True if len(pipe_input_names) == 1 else False
+            return pipe_input_names, allow_positional_args
+
+        if input_names_set != pipe_input_names_set:
+            raise RuntimeError(
+                "The set of DALI input names provided should match exactly the "
+                "ones provided by the pipeline. "
+                f"\nProvided input names are: {input_names}"
+                f"\nPipeline input names are: {pipe_input_names}"
+            )
+        return input_names, True
 
     @property
     def proxy(self):
-        return _DALIProxy(self._dali_input_names, self._dali_input_q, self._deterministic)
+        if self._proxy is None:
+            parameters = [inspect.Parameter("self", inspect.Parameter.POSITIONAL_OR_KEYWORD)]
+            for input_name in self._dali_input_names:
+                if self._allow_positional_args:
+                    parameters.append(
+                        inspect.Parameter(input_name, inspect.Parameter.POSITIONAL_OR_KEYWORD)
+                    )
+                else:
+                    parameters.append(inspect.Parameter(input_name, inspect.Parameter.KEYWORD_ONLY))
+
+            signature = inspect.Signature(parameters)
+
+            def call_impl(self, *args, **kwargs):
+                try:
+                    bound_args = signature.bind(self, *args, **kwargs)
+                except Exception as exc:
+                    raise RuntimeError(f"{exc}. Signature is {signature}")
+                return self._add_sample(bound_args)
+
+            call_impl.__signature__ = inspect.Signature(parameters)
+            _DALIProxy.__call__ = call_impl
```

Comment:
Cool 😎 
Does it have a chance of working with IDE or jupyter?

I think for IDE to work, we need the `__call__` visible statically, there is no chance of injecting a proper signature stub, so we probably still need to have a static `__call__(self, *inputs, **kwargs)` with a docstring defined, and replace it with this hook.